docs/end_to_end_tutorials/principled_dev_practices.md

```diff
@@ -0,0 +1,81 @@
+2. **Data-Driven** - How can I automate more of my evaluation and iteration process? How do I start small and improve over time? How can I manage the complexity while keeping track of my guiding light of providing the best user experience? 
+3. **Customization / Complexity Tradeoff** - Can I add additional structure. How do I improve each stage of the pipeline - preprocessing and feature extraction, retrieval, generation? How can I break down this goal into more measurable and trackable sub-goals?
+
+Differences between **Evaluation** and being **Data-Driven**:
+1. **Evaluation** is not necessarily a rigorous or full data-driven process. It is more concerned with the initial *development* phase of the application - validating that the overall pipeline and starting to define possible signals and metrics which may be carried forward when shifting gears into production.
+2. Being **Data-Driven** is closely tied to *automation*. After we've chosen our basic application structure, how can we improve the system over time? How can we ensure the quality of the system in a systematic way? How can we reduce the cost of monitoring and improving the system, and what are the pathways to adding and curating data points to add? How can we leverage ML systems to make this process easier?
+
+Additional considerations:
+1. **Privacy** - how can I ensure that my data is not leaked if I am feeding it into these models? What infrastructure am I using and what is the security guarantee / how is the access control structured?
+
+## Development Hurdles
+
+Here are some potential problems you may encounter when developing your LLM application which may lead to unsatisfactory results.
+
+### Retrieval
+
+1. **Out of Domain:**
+If your data is extremely specific (medical, legal, scientific, financial, or other documents with technical lingo), it may be worth:
+    - trying out alternate embeddings 
+      - Check the [MTEB Leaderboard](https://huggingface.co/spaces/mteb/leaderboard)
+      - You may configure a local embedding model [with the steps here](local-embedding-models)
+    - testing out fine-tuning of embeddings
+        - Tools: [setfit](https://github.com/huggingface/setfit)
+        - Anecdotally, we have seen retrieval accuracy improve by ~12% by curating a small annotated dataset from production data
+        - Even synthetic data generation without human labels has been shown to improve retrieval metrics across similar documents in train / val sets.
+        - More detailed guides and case studies will come soon.
+    - testing out sparse retrieval methods (see ColBERT, SPLADE)
+        - these methods have been shown to generalize well to out of domain data
+        - that are starting to be available in some enterprise systems (e.g. [Elastic Search's ELSeR](https://www.elastic.co/guide/en/machine-learning/current/ml-nlp-elser.html))
```

Comment:
should we link to relevant sections in evaluation just to tie the sections a bit more closely?

Reply:
Maybe we should have "pillars" and then "methods"? We can describe the methodology/pathways/case-studies for doing e.g. embedding/model fine-tuning there. Or it could be a subsection of evaluation.

Reply:
Maybe since fine-tuning is closely tied to eval, we can chuck it under eval for now.

Reply:
Hm. Yeah I think that would work if you can figure out a good subsection for it in the eval page. (i do think the concept of "finetuning"/improvement is similar to eval but separate)